zenoh-ts/src/pubsub.ts

```diff
@@ -0,0 +1,334 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+// Remote API
+import { RemoteSubscriber, RemotePublisher } from "./remote_api/pubsub";
+
+// API
+import { KeyExpr } from "./key_expr";
+import { IntoZBytes, ZBytes } from "./z_bytes";
+import {
+  CongestionControl,
+  Priority,
+  Reliability,
+  Sample,
+  Sample_from_SampleWS,
+} from "./sample";
+import { Encoding, IntoEncoding } from "./encoding";
+
+// ███████ ██    ██ ██████  ███████  ██████ ██████  ██ ██████  ███████ ██████
+// ██      ██    ██ ██   ██ ██      ██      ██   ██ ██ ██   ██ ██      ██   ██
+// ███████ ██    ██ ██████  ███████ ██      ██████  ██ ██████  █████   ██████
+//      ██ ██    ██ ██   ██      ██ ██      ██   ██ ██ ██   ██ ██      ██   ██
+// ███████  ██████  ██████  ███████  ██████ ██   ██ ██ ██████  ███████ ██   ██
+
+
+/**
+ * Class to represent a Subscriber on Zenoh, 
+ * created via calling `declare_subscriber()` on a `session`
+ */
+
+export class Subscriber {
+  /**
+   * @hidden 
+   */
+  private remote_subscriber: RemoteSubscriber;
+  /**
+   * @hidden 
+   */
+  private callback_subscriber: boolean;
+  /** Finalization registry used for cleanup on drop
+   * @hidden 
+   */
+  static registry: FinalizationRegistry<RemoteSubscriber> = new FinalizationRegistry((r_subscriber: RemoteSubscriber) => r_subscriber.undeclare());
+  /**
+   * @hidden 
+   */
+  dispose() {
+    this.undeclare();
+    Subscriber.registry.unregister(this);
+  }
+  /**
+   * @hidden 
+   */
+  constructor(
+    remote_subscriber: RemoteSubscriber,
+    callback_subscriber: boolean,
+  ) {
+    this.remote_subscriber = remote_subscriber;
+    this.callback_subscriber = callback_subscriber;
+    Subscriber.registry.register(this, remote_subscriber, this)
+  }
+
+  /**
+   * Receives a new message on the subscriber
+   *  note: If subscriber was created with a callback, this recieve will return undefined, 
+   *  as new samples are being sent to the callback.
+   *
+   * @returns Promise<Sample | void>
+   */
+  async receive(): Promise<Sample | void> {
+    if (this.callback_subscriber === true) {
+      console.log("Cannot call `receive()` on Subscriber created with callback:");
+      return;
+    }
+
+    // from SampleWS -> Sample
+    let opt_sample_ws = await this.remote_subscriber.receive();
+    if (opt_sample_ws != undefined) {
+      return Sample_from_SampleWS(opt_sample_ws);
+    } else {
+      console.log("Receieve returned unexpected void from RemoteSubscriber");
```

Comment:
```suggestion
      console.warn("Receieve returned unexpected void from RemoteSubscriber");
```

Or `console.error`.

Reply:
The same may apply to all usage of `console.log`.

Reply:
Implemented in https://github.com/eclipse-zenoh/zenoh-ts/pull/5